cli/runtime/plugin/plugin.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2021 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package plugin
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"go.uber.org/multierr"
+	"golang.org/x/mod/semver"
+
+	cliv1alpha1 "github.com/vmware-tanzu/tanzu-framework/apis/cli/v1alpha1"
+)
+
+// Plugin is a Tanzu CLI plugin.
+type Plugin struct {
+	Cmd *cobra.Command
+}
+
+// NewPlugin creates an instance of Plugin.
+func NewPlugin(descriptor *cliv1alpha1.PluginDescriptor) (*Plugin, error) {
+	ApplyDefaultConfig(descriptor)
+	err := ValidatePlugin(descriptor)
+	if err != nil {
+		return nil, err
+	}
+	p := &Plugin{
+		Cmd: newRootCmd(descriptor),
+	}
+	p.Cmd.AddCommand(lintCmd)
+	p.Cmd.AddCommand(genDocsCmd)
+	p.Cmd.AddCommand(newPostInstallCmd(descriptor))
+	return p, nil
+}
+
+// NewPluginFromFile create a new instance of Plugin from a file descriptor.
+func NewPluginFromFile(path string) (*Plugin, error) {
+	descriptor, err := parsePluginDescriptor(path)
+	if err != nil {
+		return nil, err
+	}
+	plugin, err := NewPlugin(&descriptor)
+	if err != nil {
+		return nil, err
+	}
+	return plugin, nil
+}
+
+// AddCommands adds commands to the plugin.
+func (p *Plugin) AddCommands(commands ...*cobra.Command) {
+	p.Cmd.AddCommand(commands...)
+}
+
+// Execute executes the plugin.
+func (p *Plugin) Execute() error {
+	return p.Cmd.Execute()
+}
+
+// parsePluginDescriptor parses a plugin descriptor in yaml.
+func parsePluginDescriptor(path string) (desc cliv1alpha1.PluginDescriptor, err error) {
+	b, err := os.ReadFile(path)
+	if err != nil {
+		return desc, errors.Wrap(err, "could not read plugin descriptor")
+	}
+
+	err = json.Unmarshal(b, &desc)
+	if err != nil {
+		return desc, errors.Wrap(err, "could not unmarshal plugin descriptor")
+	}
+	return
+}
+
+// ApplyDefaultConfig applies default configurations to plugin descriptor.
+func ApplyDefaultConfig(p *cliv1alpha1.PluginDescriptor) {
+	// TODO(anuj): Removed next lines as buildsha and version is not relevant to library and needs to be configured per plugin
+	// p.BuildSHA = buildinfo.SHA
```

Comment:
Since we won't be setting the sha automatically, what do you think about adding it in the template used to generate plugin code?
https://github.com/vmware-tanzu/tanzu-framework/blob/0605dfc98500015c474ca3709e78d8cb434ac607/pkg/v1/builder/template/plugintemplates/main.go.tmpl#L12-L17

Reply:
Yes. Makes sense